### PR TITLE
Consistent hash bounded load

### PR DESCRIPTION
**Pull request by herojan**
Implements https://github.com/zalando/skipper/issues/1769

### CODE REVIEW

File: eskipfile/watch_test.go [hunk 1]
```diff
@@ -49,19 +51,42 @@ func createFileWith(content string, t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	defer os.Remove(tmpfile.Name())
```
Comment (szuecs) [inline eskipfile/watch_test.go:52]: removing this is a file descriptor leak

File: eskipfile/watch_test.go [hunk 2]
```diff
@@ -49,19 +51,42 @@ func createFileWith(content string, t *testing.T) {
...
 
 	_, err = tmpfile.WriteString(content)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	err = os.Rename(tmpfile.Name(), testWatchFile)
+	err = moveFile(tmpfile.Name(), testWatchFile)
```
Comment (szuecs) [inline eskipfile/watch_test.go:60]: why not pass the tmpFile, which is already open?

Comment (herojan) [reply eskipfile/watch_test.go:60]: Yeah I copy pasted this function as the first thing I did to get tests passing, I'll do as you suggest and see if it can be cleaned up a bit before taking the pr out of draft

File: eskipfile/watch_test.go (hunk 1 above)
Comment (herojan) [reply eskipfile/watch_test.go:52]: Is it still a leak if I call os.Remove in the moveFile function?

Comment (herojan) [reply eskipfile/watch_test.go:52]: Ah I see, if something fails in moveFile then Remove won't be called :+1:

File: loadbalancer/algorithm.go [hunk 3]
```diff
@@ -220,14 +221,56 @@ func hash(s string) uint32 {
...
+func computeLoadAverage(ctx *routing.LBContext) float64 {
+	sum := 1.0 // add 1 to include the request that just arrived
+	endpoints := ctx.Route.LBEndpoints
+	numEndpoints := len(ctx.Route.LBEndpoints)
+	for _, v := range endpoints {
+		sum += float64(v.Metrics.GetInflightRequests())
+	}
+	if numEndpoints == 0 {
+		return 0.0
+	} else {
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:250]: We ensure earlier there at least two endpoints

File: loadbalancer/algorithm.go [hunk 4]
```diff
@@ -220,14 +221,56 @@ func hash(s string) uint32 {
...
+	}
+}
+
+// Returns index of endpoint with closest hash to key's hash, which is also below the target load
+func (ch consistentHash) boundedLoadSearch(key string, balanceFactor float64, ctx *routing.LBContext) int {
+	ringIndex := ch.searchRing(key)
+	averageLoad := computeLoadAverage(ctx)
+	targetLoad := averageLoad * balanceFactor
+	// Loop round ring, starting at endpoint with closest hash. Stop when we find one whose load is less than targetLoad.
+	ringIndex -= 1
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:261]: No need to adjust it if we move
```go
ringIndex = (ringIndex + 1) % ch.Len()
```
to the end of the loop body

File: loadbalancer/algorithm.go [hunk 5]
```diff
@@ -220,14 +221,56 @@ func hash(s string) uint32 {
...
+	for i := 0; i < ch.Len(); i++ {
+		ringIndex = (ringIndex + 1) % ch.Len()
+		endpointIndex := ch[ringIndex].index
+		inflightRequests := ctx.Route.LBEndpoints[endpointIndex].Metrics.GetInflightRequests()
+		// We know there must be an endpoint whose load <= average load.
+		// Since targetLoad >= average load (balancerFactor >= 1), there must also be an endpoint with load <= targetLoad.
+		if inflightRequests <= int(targetLoad) {
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:268]: Maybe cast `inflightRequests` to `float64` instead?

Comment (AlexanderYastrebov) [reply loadbalancer/algorithm.go:268]:
```go
ep := ctx.Route.LBEndpoints[endpointIndex]
load := float64(ep.Metrics.GetInflightRequests())
```

File: loadbalancer/algorithm.go [hunk 6]
```diff
@@ -220,14 +221,56 @@ func hash(s string) uint32 {
...
+	ringIndex := ch.searchRing(key)
+	averageLoad := computeLoadAverage(ctx)
+	targetLoad := averageLoad * balanceFactor
+	// Loop round ring, starting at endpoint with closest hash. Stop when we find one whose load is less than targetLoad.
+	ringIndex -= 1
+	for i := 0; i < ch.Len(); i++ {
+		ringIndex = (ringIndex + 1) % ch.Len()
+		endpointIndex := ch[ringIndex].index
+		inflightRequests := ctx.Route.LBEndpoints[endpointIndex].Metrics.GetInflightRequests()
+		// We know there must be an endpoint whose load <= average load.
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:266]: Metrics may change during iteration hence we use finite loop. 
If we were certain we could use
```go
for float64(ctx.Route.LBEndpoints[ch[ringIndex].index].Metrics.GetInflightRequests()) > targetLoad {
    ringIndex = (ringIndex + 1) % len(ch)
}
```

File: loadbalancer/algorithm.go [hunk 7]
```diff
@@ -220,14 +221,56 @@ func hash(s string) uint32 {
...
+}
+
+// Returns index of endpoint with closest hash to key's hash
+func (ch consistentHash) search(key string) int {
+	ringIndex := ch.searchRing(key)
+	return ch[ringIndex].index
+}
+
+func computeLoadAverage(ctx *routing.LBContext) float64 {
+	sum := 1.0 // add 1 to include the request that just arrived
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:242]: Why do we need to start with 1?

Comment (AlexanderYastrebov) [reply loadbalancer/algorithm.go:242]: I.e. algo is looking for backend to forward this request why would it count it into the load?

File: loadbalancer/algorithm_test.go [hunk 8]
```diff
@@ -312,6 +312,58 @@ func TestConsistentHashSearch(t *testing.T) {
...
+	r.RemoteAddr = "192.168.0.1:8765"
+	route := NewAlgorithmProvider().Do([]*routing.Route{{
+		Route: eskip.Route{
+			BackendType: eskip.LBBackend,
+			LBAlgorithm: ConsistentHash.String(),
+			LBEndpoints: endpoints,
+		},
+	}})[0]
+	ctx := &routing.LBContext{Request: r, Route: route, Params: map[string]interface{}{ConsistentHashKey: net.RemoteHost(r).String()}}
+	noLoad := ch.boundedLoadSearch("somekey", 1.25, ctx)
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm_test.go:329]: Should we rather test using `ch.Apply(&routing.LBContext{Request: r, Route: rt, Params: map[string]interface{...}})`?

To compare bounded with non bounded we may call once with `ConsistentHashBalanceFactor: 1.25` param and once without.

Comment (AlexanderYastrebov) [reply loadbalancer/algorithm_test.go:329]: Maybe drop `ConsistentHashKey` param for clarity - it is not relevant to the bounded algo

File: loadbalancer/algorithm_test.go [hunk 9]
```diff
@@ -312,6 +312,58 @@ func TestConsistentHashSearch(t *testing.T) {
...
+	nonBounded := ch.search("somekey")
+
+	if noLoad != nonBounded {
+		t.Error("When no endpoints are overloaded, the chosen endpoint should be the same as standard consistentHash")
+	}
+	// now we know that noLoad is the endpoint which should be requested for somekey if load is not an issue.
+	for range [20]int{} {
+		route.LBEndpoints[noLoad].Metrics.IncInflightRequest()
+	}
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm_test.go:338]: Should we introduce a reusbale helper function for this like
```go
addInflightRequests(route.LBEndpoints[noLoad], 20)
```

File: loadbalancer/algorithm.go [hunk 10]
```diff
@@ -220,14 +221,56 @@ func hash(s string) uint32 {
 	return h.Sum32()
 }
 
-// Returns index of endpoint with closest hash to key's hash
-func (ch consistentHash) search(key string) int {
+// returns index in hash ring with the closest hash to key's hash
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:224]: `Returns` from capital for consistency

File: loadbalancer/algorithm.go [hunk 11]
```diff
@@ -220,14 +221,56 @@ func hash(s string) uint32 {
...
+func (ch consistentHash) searchRing(key string) int {
 	h := hash(key)
 	i := sort.Search(ch.Len(), func(i int) bool { return ch[i].hash >= h })
 	if i == ch.Len() { // rollover
 		i = 0
 	}
-	return ch[i].index
+
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:231]: IMO no need for new line here

File: eskipfile/watch_test.go [hunk 12]
```diff
@@ -45,18 +44,12 @@ func deleteFile(t *testing.T) {
 }
 
 func createFileWith(content string, t *testing.T) {
-	tmpfile, err := ioutil.TempFile("", "watch_test-*")
```
Comment (AlexanderYastrebov) [inline eskipfile/watch_test.go:48]: Should we do this changes in separate PR? Would be easier to dig git history down the line

File: loadbalancer/algorithm.go (hunk 5 above)
Comment (herojan) [reply loadbalancer/algorithm.go:268]: In the paper it said to pick floor or ceiling of targetLoad so I cast to int to get the floor in this case. I'll switch the naming to load from inflightRequests though

File: loadbalancer/algorithm.go (hunk 6 above)
Comment (herojan) [reply loadbalancer/algorithm.go:266]: Yeah probably best to not include new endpoints in the calculation as they arrive. That loop looked very confusing for a second until I realised it's basically a while loop :smiley:

File: loadbalancer/algorithm.go (hunk 7 above)
Comment (herojan) [reply loadbalancer/algorithm.go:242]: The paper said to include it, but the reason is that if you look at the load of the existing servers at any time it should be true that no server is overloaded by more than `(targetLoad*balanceFactor) +1` (+1 since we took the floor of targetLoad). If we don't include the incoming request, then that property may not hold after the incoming request is added to a backend server.

File: loadbalancer/algorithm_test.go (hunk 8 above)
Comment (herojan) [reply loadbalancer/algorithm_test.go:329]: Yeah true I'm missing a new if branch if I don't test Apply, I'll switch over to that

File: loadbalancer/algorithm_test.go (hunk 9 above)
Comment (herojan) [reply loadbalancer/algorithm_test.go:338]: Is there a convention on where to put helper functions for tests? Or can I just put it as a private method in algorithm_test.go

File: eskipfile/watch_test.go (hunk 12 above)
Comment (herojan) [reply eskipfile/watch_test.go:48]: Ok, I'll split it out.

Comment (herojan) [reply eskipfile/watch_test.go:48]: https://github.com/zalando/skipper/pull/1799 Split it out to here

File: loadbalancer/algorithm.go (hunk 6 above)
Comment (AlexanderYastrebov) [reply loadbalancer/algorithm.go:259]: > Yeah probably best to not include new endpoints in the calculation as they arrive

endpoint list does not change, but inflight request count of an endpoint might change during iteration (and avg load calculation as well)

File: loadbalancer/algorithm_test.go [hunk 13]
```diff
@@ -312,6 +312,46 @@ func TestConsistentHashSearch(t *testing.T) {
...
+	ch := route.LBAlgorithm.(consistentHash)
+	ctx := &routing.LBContext{Request: r, Route: route, Params: map[string]interface{}{ConsistentHashBalanceFactor: 1.25}}
+	noLoad := ch.Apply(ctx)
+	nonBounded := ch.Apply(&routing.LBContext{Request: r, Route: route, Params: map[string]interface{}{}})
+
+	if noLoad != nonBounded {
+		t.Error("When no endpoints are overloaded, the chosen endpoint should be the same as standard consistentHash")
+	}
+	// now we know that noLoad is the endpoint which should be requested for somekey if load is not an issue.
+	addInflightRequests(noLoad, 20)
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm_test.go:334]: After some thought I think we have a problem here. It could not happen that single backend gets +20 inflight requests when requests are balanced by the algorithm because algorithm would prevent them from going to the same backend (avg load = 20/3 = 6.6, max allowed load is 1.25*6.6=8.33 which is less than 20)

I think we need to change test approach and e.g. issue N requests (with a same key) and check that
* each and every backend is not overloaded 
* (optional) spillover happens along the ring (increasing hash)

Another test may issue a set of requests with different hash keys but again no backend should be overloaded.

Comment (AlexanderYastrebov) [reply loadbalancer/algorithm_test.go:334]:
```go
func TestConsistentHashBoundedLoadDistribution(t *testing.T) {
	endpoints := []string{"http://127.0.0.1:8080", "http://127.0.0.2:8080", "http://127.0.0.3:8080"}
	r, _ := http.NewRequest("GET", "http://127.0.0.1:1234/foo", nil)
	route := NewAlgorithmProvider().Do([]*routing.Route{{
		Route: eskip.Route{
			BackendType: eskip.LBBackend,
			LBAlgorithm: ConsistentHash.String(),
			LBEndpoints: endpoints,
		},
	}})[0]
	ch := route.LBAlgorithm.(consistentHash)
	ctx := &routing.LBContext{Request: r, Route: route, Params: map[string]interface{}{ConsistentHashBalanceFactor: 1.25}}

	for i := 0; i < 100; i++ {
		ep := ch.Apply(ctx)
		ep.Metrics.IncInflightRequest() // done by proxy

		t.Logf("%d %d %d",
			route.LBEndpoints[ch[0].index].Metrics.GetInflightRequests(),
			route.LBEndpoints[ch[1].index].Metrics.GetInflightRequests(),
			route.LBEndpoints[ch[2].index].Metrics.GetInflightRequests())
	}
```
```
go test -run TestConsistentHashBoundedLoadDistribution ./loadbalancer -v
```
```
=== RUN   TestConsistentHashBoundedLoadDistribution
    algorithm_test.go:409: 1 0 0
    algorithm_test.go:409: 1 1 0
    algorithm_test.go:409: 2 1 0
    algorithm_test.go:409: 2 2 0
    algorithm_test.go:409: 3 2 0
    algorithm_test.go:409: 3 3 0
    algorithm_test.go:409: 3 3 1
    algorithm_test.go:409: 4 3 1
    algorithm_test.go:409: 4 4 1
    algorithm_test.go:409: 5 4 1
    algorithm_test.go:409: 5 5 1
    algorithm_test.go:409: 6 5 1
    algorithm_test.go:409: 6 6 1
    algorithm_test.go:409: 6 6 2
    algorithm_test.go:409: 7 6 2
    algorithm_test.go:409: 7 7 2
    algorithm_test.go:409: 8 7 2
    algorithm_test.go:409: 8 8 2
    algorithm_test.go:409: 8 8 3
    algorithm_test.go:409: 9 8 3
    algorithm_test.go:409: 9 9 3
    algorithm_test.go:409: 10 9 3
    algorithm_test.go:409: 10 10 3
    algorithm_test.go:409: 11 10 3
    algorithm_test.go:409: 11 11 3
    algorithm_test.go:409: 11 11 4
    algorithm_test.go:409: 12 11 4
    algorithm_test.go:409: 12 12 4
    algorithm_test.go:409: 13 12 4
    algorithm_test.go:409: 13 13 4
    algorithm_test.go:409: 13 13 5
    algorithm_test.go:409: 14 13 5
    algorithm_test.go:409: 14 14 5
    algorithm_test.go:409: 15 14 5
    algorithm_test.go:409: 15 15 5
    algorithm_test.go:409: 16 15 5
    algorithm_test.go:409: 16 16 5
    algorithm_test.go:409: 16 16 6
    algorithm_test.go:409: 17 16 6
    algorithm_test.go:409: 17 17 6
    algorithm_test.go:409: 18 17 6
    algorithm_test.go:409: 18 18 6
    algorithm_test.go:409: 18 18 7
    algorithm_test.go:409: 19 18 7
    algorithm_test.go:409: 19 19 7
    algorithm_test.go:409: 20 19 7
    algorithm_test.go:409: 20 20 7
    algorithm_test.go:409: 21 20 7
    algorithm_test.go:409: 21 21 7
    algorithm_test.go:409: 21 21 8
    algorithm_test.go:409: 22 21 8
    algorithm_test.go:409: 22 22 8
    algorithm_test.go:409: 23 22 8
    algorithm_test.go:409: 23 23 8
    algorithm_test.go:409: 23 23 9
    algorithm_test.go:409: 24 23 9
    algorithm_test.go:409: 24 24 9
    algorithm_test.go:409: 25 24 9
    algorithm_test.go:409: 25 25 9
    algorithm_test.go:409: 26 25 9
    algorithm_test.go:409: 26 26 9
    algorithm_test.go:409: 26 26 10
    algorithm_test.go:409: 27 26 10
    algorithm_test.go:409: 27 27 10
    algorithm_test.go:409: 28 27 10
    algorithm_test.go:409: 28 28 10
    algorithm_test.go:409: 28 28 11
    algorithm_test.go:409: 29 28 11
    algorithm_test.go:409: 29 29 11
    algorithm_test.go:409: 30 29 11
    algorithm_test.go:409: 30 30 11
    algorithm_test.go:409: 31 30 11
    algorithm_test.go:409: 31 31 11
    algorithm_test.go:409: 31 31 12
    algorithm_test.go:409: 32 31 12
    algorithm_test.go:409: 32 32 12
    algorithm_test.go:409: 33 32 12
    algorithm_test.go:409: 33 33 12
    algorithm_test.go:409: 33 33 13
    algorithm_test.go:409: 34 33 13
    algorithm_test.go:409: 34 34 13
    algorithm_test.go:409: 35 34 13
    algorithm_test.go:409: 35 35 13
    algorithm_test.go:409: 36 35 13
    algorithm_test.go:409: 36 36 13
    algorithm_test.go:409: 36 36 14
    algorithm_test.go:409: 37 36 14
    algorithm_test.go:409: 37 37 14
    algorithm_test.go:409: 38 37 14
    algorithm_test.go:409: 38 38 14
    algorithm_test.go:409: 38 38 15
    algorithm_test.go:409: 39 38 15
    algorithm_test.go:409: 39 39 15
    algorithm_test.go:409: 40 39 15
    algorithm_test.go:409: 40 40 15
    algorithm_test.go:409: 41 40 15
    algorithm_test.go:409: 41 41 15
    algorithm_test.go:409: 41 41 16
    algorithm_test.go:409: 42 41 16
    algorithm_test.go:409: 42 42 16
```

Comment (herojan) [reply loadbalancer/algorithm_test.go:334]: > It could not happen that single backend gets +20 inflight requests when requests are balanced by the algorithm

Yes that can't happen, I just did it for the test so that it is obvious to the reader which one is overloaded / above average. I can add the test you show above though to show no server is overloaded as requests come in :+1:

File: filters/consistenthash/balancefactor.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+type consistentHashBalanceFactor struct {
+	balanceFactor float64
+}
+
+// NewConsistentHashBalanceFactor creates a filter Spec, whose instances
+// set the balancer factor used by the `consistentHash` algorithm to avoid
+// popular hashes overloading a single endpoint
+func NewConsistentHashBalanceFactor() filters.Spec { return &consistentHashBalanceFactor{} }
```
Comment (AlexanderYastrebov) [inline filters/consistenthash/balancefactor.go:17]: We should not forget to add user docs similar to https://github.com/zalando/skipper/blob/master/docs/reference/filters.md#consistenthashkey

File: docs/reference/filters.md [hunk 15]
```diff
@@ -2385,3 +2385,27 @@ pr: Path("/products/:productId")
 consistentHashKey("${request.header.Authorization}")
 consistentHashKey("${request.source}") // same as the default key
 ```
+
+## consistentHashBalanceFactor
+
+This filter sets the balance factor used to prevent a single backend endpoint from being overloaded.
+The number of in-flight requests for an endpoint can be no higher than `(average-in-flight-requests * balanceFactor) + 1`.
+This is helpful in the case where certain keys are very popular and threaten to overload the endpoint they are mapped to.
+[Further Details](https://ai.googleblog.com/2017/04/consistent-hashing-with-bounded-loads.html).
```
Comment (AlexanderYastrebov) [inline docs/reference/filters.md:2394]: Additionally it may also link to https://github.com/zalando/skipper/blob/master/docs/reference/backends.md#load-balancer-backend and that page may link back to this filter like it does for `consistentHashKey`

File: docs/reference/filters.md [hunk 16]
```diff
@@ -2437,3 +2437,27 @@ pr: Path("/products/:productId")
...
+## consistentHashBalanceFactor
+
+This filter sets the balance factor used by the [`consistentHash`](backends.md#load-balancer-backend) algorithm to prevent a single backend endpoint from being overloaded.
+The number of in-flight requests for an endpoint can be no higher than `(average-in-flight-requests * balanceFactor) + 1`.
+This is helpful in the case where certain keys are very popular and threaten to overload the endpoint they are mapped to.
+[Further Details](https://ai.googleblog.com/2017/04/consistent-hashing-with-bounded-loads.html).
+
+Parameters:
+
+* balanceFactor (float)
```
Comment (AlexanderYastrebov) [inline docs/reference/filters.md:2450]: Lets maybe mention that it must be greater or equal to 1?

File: docs/reference/filters.md [hunk 17]
```diff
@@ -2437,3 +2437,27 @@ pr: Path("/products/:productId")
...
+
+```
+pr: Path("/products/:productId")
+    -> consistentHashKey("${productId}")
+    -> consistentHashBalanceFactor(1.25)
+    -> <consistentHash, "http://127.0.0.1:9998", "http://127.0.0.1:9997">;
+```
+```
+consistentHashBalanceFactor(1.55)
+consistentHashBalanceFactor(1)
```
Comment (AlexanderYastrebov) [inline docs/reference/filters.md:2462]: IMO these are redundant

File: filters/consistenthash/balancefactor_test.go [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+	f, _ := spec.CreateFilter([]interface{}{1.0})
+	f.Request(c)
+
+	if c.FStateBag[loadbalancer.ConsistentHashBalanceFactor] != 1.0 {
+		t.Error("Failed to set balance factor via filter")
+	}
+
+	_, e := spec.CreateFilter([]interface{}{0.5})
+
+	if e == nil {
```
Comment (AlexanderYastrebov) [inline filters/consistenthash/balancefactor_test.go:29]: minor: `err`

File: loadbalancer/algorithm.go [hunk 19]
```diff
@@ -39,6 +39,7 @@ const (
 const powerOfRandomNChoicesDefaultN = 2
 
 const ConsistentHashKey = "consistentHashKey"
+const ConsistentHashBalanceFactor = "consistentHashBalanceFactor"
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:42]: minor: could be also grouped like
```go
const (
    ConsistentHashKey = "consistentHashKey"
    ConsistentHashBalanceFactor = "consistentHashBalanceFactor"
)
```

File: loadbalancer/algorithm.go [hunk 20]
```diff
@@ -220,14 +221,50 @@ func hash(s string) uint32 {
...
+// Returns index of endpoint with closest hash to key's hash
+func (ch consistentHash) search(key string) int {
+	ringIndex := ch.searchRing(key)
+	return ch[ringIndex].index
+}
+
+func computeLoadAverage(ctx *routing.LBContext) float64 {
+	sum := 1.0 // add 1 to include the request that just arrived
+	endpoints := ctx.Route.LBEndpoints
+	numEndpoints := len(ctx.Route.LBEndpoints)
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:243]: minor: lets inline this maybe?
```go
return sum / float64(len(endpoints))
```

File: docs/reference/filters.md [hunk 21]
```diff
@@ -2437,3 +2437,27 @@ pr: Path("/products/:productId")
 consistentHashKey("${request.header.Authorization}")
 consistentHashKey("${request.source}") // same as the default key
 ```
+
+## consistentHashBalanceFactor
+
+This filter sets the balance factor used by the [`consistentHash`](backends.md#load-balancer-backend) algorithm to prevent a single backend endpoint from being overloaded.
+The number of in-flight requests for an endpoint can be no higher than `(average-in-flight-requests * balanceFactor) + 1`.
```
Comment (AlexanderYastrebov) [inline docs/reference/filters.md:2444]: I am not sure where does `+ 1` comes from, the impl uses 
```go
targetLoad := averageLoad * balanceFactor
...
if load <= int(targetLoad) {
    break
}
``` 
Lets remove and simplfy?

**Comment by AlexanderYastrebov** [comment]
@herojan Thank you for your efforts and contribution

**Comment by AlexanderYastrebov** [comment]
:+1:

File: docs/reference/filters.md (hunk 21 above)
Comment (herojan) [reply docs/reference/filters.md:2444]: After we assign a request to the server the number of in-flight requests will be targetLoad + 1

File: docs/reference/filters.md (hunk 17 above)
Comment (herojan) [reply docs/reference/filters.md:2462]: With the examples I was trying to show that you can use an int or a float as the value

**Comment by herojan** [comment]
Thanks for the thorough reviews! @AlexanderYastrebov

**Comment by AlexanderYastrebov** [comment]
:+1:

**Comment by aryszka** [comment]
This is great! Thank you @herojan and @AlexanderYastrebov !

**Comment by aryszka** [comment]
:+1:
